homeassistant/components/goodwe/sensor.py

```diff
@@ -0,0 +1,257 @@
+
+    _MAIN_ENTITY_SENSOR = "ppv"
+
+    def __init__(self, coordinator, device_info, inverter):
+        """Initialize the main inverter entity."""
+        super().__init__(coordinator)
+        self.entity_id = f".{DOMAIN}_inverter"
+        self._attr_unique_id = f"{DOMAIN}-{inverter.serial_number}"
+        self._attr_device_info = device_info
+        self._attr_icon = "mdi:solar-power"
+        self._attr_name = "PV Inverter"
+        self._attr_native_unit_of_measurement = POWER_WATT
+        self._inverter = inverter
+
+    async def set_work_mode(self, work_mode: int):
+        """Set the inverter work mode."""
+        await self._inverter.set_work_mode(work_mode)
+
+    async def set_ongrid_battery_dod(self, ongrid_battery_dod: int):
+        """Set the on-grid battery dod."""
+        await self._inverter.set_ongrid_battery_dod(ongrid_battery_dod)
+
+    async def set_grid_export_limit(self, grid_export_limit: int):
+        """Set the grid export limit."""
+        await self._inverter.set_grid_export_limit(grid_export_limit)
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self.coordinator.data is not None
+
+    @property
+    def native_value(self):
+        """Return the value reported by the sensor."""
+        if self.coordinator.data is not None:
+            new_value = self.coordinator.data.get(self._MAIN_ENTITY_SENSOR)
+            # If no new value was provided, keep the previous
+            if new_value is not None:
+                self._attr_native_value = new_value
+
+        return self._attr_native_value
+
+    @property
+    def state_attributes(self):
+        """Return the inverter state attributes."""
+        return self.coordinator.data
+
+    @property
+    def extra_state_attributes(self):
+        """Return the inverter state attributes."""
+        data = {
+            "model": self._inverter.model_name,
+            "serial_number": self._inverter.serial_number,
+            "firmware_version": self._inverter.software_version,
+            "arm_version": self._inverter.arm_version,
+        }
+        return data
+
+
+class InverterSensor(CoordinatorEntity, SensorEntity):
+    """Entity representing individual inverter sensor."""
+
+    def __init__(self, coordinator, device_info, description, inverter):
+        """Initialize an inverter sensor."""
+        super().__init__(coordinator)
+        self.entity_description = description
+        self.entity_id = f".{DOMAIN}_{description.key}"
+        self._attr_unique_id = f"{DOMAIN}-{description.key}-{inverter.serial_number}"
+        self._attr_device_info = device_info
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self.coordinator.data is not None
+
+    @property
+    def native_value(self):
+        """Return the value reported by the sensor."""
+        if self.coordinator.data is not None:
```

Comment:
This check certainly doesn't hurt. I don't know the update coordinator that well so I don't know if it is necessary. Leave it as it unless you discover its not needed or someone else tells you...

Reply:
removed as consequence above